File: src/main/java/de/take_weiland/mods/commons/util/NBT.java
package de.take_weiland.mods.commons.util;

import java.util.List;

import net.minecraft.nbt.NBTBase;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraft.nbt.NBTTagString;

import com.google.common.base.Function;

public final class NBT {

	private NBT() { }

	/**
	 * view the given NBTTagList as a {@link List}<br>
	 * the type parameter T can be used if you are sure that this list only contains NBT-Tags of the given type
	 * @param nbtList the list to view
	 * @return a modifiable list view of the NBTTagList
	 */
	public static <T extends NBTBase> List<T> asList(NBTTagList nbtList) {
		return MiscUtil.getReflector().getWrappedList(nbtList);
	}

	public static NBTTagCompound getOrCreateCompound(NBTTagCompound parent, String key) {
		if (!parent.hasKey(key)) {
			parent.setCompoundTag(key, new NBTTagCompound());
		}
		return parent.getCompoundTag(key);
	}
	
	public static NBTTagList getOrCreateList(NBTTagCompound parent, String key) {
		if (!parent.hasKey(key)) {
			parent.setTag(key, new NBTTagList());
		}
		return parent.getTagList(key);
	}
	
	private static enum NbtStringDataFunction implements Function<NBTTagString, String> {
		
		INSTANCE;
		
		@Override
		public String apply(NBTTagString input) {
			return input.data;
		}
		
	}

	public static Function<NBTTagString, String> getStringFunction() {
		return NbtStringDataFunction.INSTANCE;
	}
	
}

File: src/main/kotlin/de/take_weiland/mods/commons/util/nbt.kt
package de.take_weiland.mods.commons.util

import net.minecraft.nbt.NBTBase
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraft.nbt.NBTTagString

/**
 * @author diesieben07
 */
inline fun NBTTagList.forEach(action: (NBTBase) -> Unit) {
    this.forEach<NBTBase> { action(it) }
}

@JvmName("forEachTyped")
inline fun <reified T : NBTBase> NBTTagList.forEach(action: (T) -> Unit) {
    tagCount().let {
        var i = 0
        while (i < it) {
            this[i].let {
                if (it is T) action(it) else throw IllegalStateException("NBTTagList contains unexpected tag $it, expected ${T::class.java}")
            }
            i++
        }
    }
}

@JvmName("forEachString")
inline fun NBTTagList.forEach(action: (String) -> Unit) {
    forEach<NBTTagString> { action(it.string) }
}

operator fun NBTTagList.plusAssign(tag: NBTBase) = appendTag(tag)
operator fun NBTTagList.iterator(): MutableIterator<NBTBase> = tagList.iterator()

operator fun NBTTagList.plusAssign(value: String) {
    appendTag(NBTTagString(value))
}

inline fun NBTTagCompound.forEach(crossinline action: (String, NBTBase) -> Unit) {
    tagMap.forEach({ k, v -> action(k, v) })
}
